src/main/java/mods/eln/misc/FloatLinear.kt

```diff
@@ -0,0 +1,130 @@
+package mods.eln.misc
+
+data class Vec3f(var x: Float, var y: Float, var z: Float) {
+    companion object {
+        val ZERO = Vec3f(0.0f, 0.0f, 0.0f)
+        val ONE = Vec3f(1.0f, 1.0f, 1.0f)
+        val AXIS_X = Vec3f(1.0f, 0.0f, 0.0f)
+        val AXIS_Y = Vec3f(0.0f, 1.0f, 0.0f)
+        val AXIS_Z = Vec3f(0.0f, 0.0f, 1.0f)
+        val AXES = arrayOf(AXIS_X, AXIS_Y, AXIS_Z)
+
+        fun fromVolumeCenter(c: Coordonate) = Vec3f(c.x + 0.5f, c.y + 0.5f, c.z + 0.5f)
+    }
+
+    constructor(c: Coordonate) : this(c.x.toFloat(), c.y.toFloat(), c.z.toFloat())
+    fun toCoordonate(dim: Int = 0) = Coordonate(
+        Math.floor(x.toDouble()).toInt(),
+        Math.floor(y.toDouble()).toInt(),
+        Math.floor(z.toDouble()).toInt(),
+        dim
+    )
+
+    fun dot(other: Vec3f) = x * other.x + y * other.y + z * other.z
+
+    fun negate(): Vec3f = Vec3f(-x, -y, -z)
+    operator fun unaryMinus() = negate()
+    fun reciprocal(): Vec3f = Vec3f(1/x, 1/y, 1/z)
+    fun add(other: Vec3f) = Vec3f(x + other.x, y + other.y, z + other.z)
+    fun add(other: Float) = Vec3f(x + other, y + other, z + other)
+    operator fun plus(other: Vec3f) = add(other)
+    operator fun plus(other: Float) = add(other)
+    fun sub(other: Vec3f) = add(other.negate())
+    fun sub(other: Float) = add(-other)
+    operator fun minus(other: Vec3f) = sub(other)
+    operator fun minus(other: Float) = sub(other)
+    fun mul(other: Vec3f) = Vec3f(x * other.x, y * other.y, z * other.z)
+    fun mul(other: Float) = Vec3f(x * other, y * other, z * other)
+    operator fun times(other: Vec3f) = mul(other)
+    operator fun times(other: Float) = mul(other)
+    operator fun div(other: Vec3f) = mul(other.reciprocal())
+    operator fun div(other: Float) = mul(1/other)
+
+    fun magSquared() = dot(this)
+    fun mag() = Math.sqrt(magSquared().toDouble()).toFloat()
+    fun isZero() = magSquared() == 0.0f
+
+    fun distanceTo(other: Vec3f) = sub(other).mag()
+    fun normalized() = if(isZero()) Vec3f(0.0f, 0.0f, 0.0f) else div(mag())
+}
+
+data class Line3f(var origin: Vec3f, var vector: Vec3f) {
+    companion object {
+        fun fromEndpoints(a: Vec3f, b: Vec3f) = Line3f(a, b.sub(a))
+    }
+
+    fun pointAt(u: Float) = origin.add(vector.mul(u))
+    fun midpoint() = pointAt(0.5f)
+    fun opposite() = pointAt(1.0f)
+    fun length() = vector.mag()
+
+    fun allIntersectionsAlongU(): List<Float> {
```

Comment:
If you got this algorithm from somewhere, please include a link. That goes for all of this.